src/loaders/lottie/tvgLottieBuilder.cpp

```diff
@@ -1224,63 +1224,65 @@ void LottieBuilder::updateMaskings(LottieLayer* layer, float frameNo)
-        layer->scene->clip(pShape);
-        return;
-    }
-
-    //Introduce an intermediate scene for embracing the matte + masking
-    if (layer->matteTarget) {
-        auto scene = Scene::gen();
-        scene->push(layer->scene);
-        layer->scene = scene;
-    }
+    Shape* pShape = nullptr;
+    MaskMethod pMethod;
+    uint8_t pOpacity;
 
-    layer->scene->mask(pShape, compMethod);
+    //find the first valid mask
+    for (auto m = layer->masks.begin(); m < layer->masks.end(); ++m) {
+        auto mask = *m;
+        if (mask->method == MaskMethod::None) continue;
 
-    //Apply the subsquent masks
-    for (auto m = layer->masks.begin() + 1; m < layer->masks.end(); ++m) {
-        auto mask = static_cast<LottieMask*>(*m);
         auto method = mask->method;
-        if (method == MaskMethod::None) continue;
-
-        //Append the mask shape
-        if (pMethod == method && (method == MaskMethod::Subtract || method == MaskMethod::Difference)) {
-            mask->pathset(frameNo, SHAPE(pShape)->rs.path.cmds, SHAPE(pShape)->rs.path.pts, nullptr, nullptr, nullptr, exps);
+        auto opacity = mask->opacity(frameNo);
+        auto expand = mask->expand(frameNo);
+        auto fastTrack = false;  //single clipping
+
+        if (!pShape) {
+            pShape = layer->pooling();
+            pShape->reset();
+            pShape->fill(255, 255, 255, opacity);
+            pShape->transform(layer->cache.matrix);
+            auto compMethod = (method == MaskMethod::Subtract || method == MaskMethod::InvAlpha) ? MaskMethod::InvAlpha : MaskMethod::Alpha;
+            //Cheaper. Replace the masking with a clipper
+            if (layer->masks.count == 1 && compMethod == MaskMethod::Alpha && opacity == 255) {
+                layer->scene->clip(pShape);
+                fastTrack = true;
+            } else {
+                layer->scene->mask(pShape, compMethod);
+            }
         //Chain composition
-        } else {
+        } else if (pMethod != method || pOpacity != opacity || (method != MaskMethod::Subtract && method != MaskMethod::Difference)) {
             auto shape = layer->pooling();
             shape->reset();
-            shape->fill(255, 255, 255, mask->opacity(frameNo));
+            pOpacity = mask->opacity(frameNo);
+            shape->fill(255, 255, 255, pOpacity);
             shape->transform(layer->cache.matrix);
-            mask->pathset(frameNo, SHAPE(shape)->rs.path.cmds, SHAPE(shape)->rs.path.pts, nullptr, nullptr, nullptr, exps);
             pShape->mask(shape, method);
             pShape = shape;
-            pMethod = method;
         }
+
+        //Default Masking
+        if (expand == 0.0f) {
+            mask->pathset(frameNo, SHAPE(pShape)->rs.path.cmds, SHAPE(pShape)->rs.path.pts, nullptr, nullptr, nullptr, exps);
+        //Masking with Expansion (Offset)
+        } else {
+            //TODO: Once path direction support is implemented, ensure that the direction is ignored here
```

Comment:
@mgrudzinska please confirm this, the offset could be applied to all subsequent masking shapes. (before you only applied it to the first masking shape)

Reply:
exactly the same :)
before only one of the masks was expanding

AE:
![multiMask_AE](https://github.com/user-attachments/assets/2fa6800a-61d0-4e21-81de-84a4dffcae40)

PR:
![multimask_PR](https://github.com/user-attachments/assets/5d52f856-674b-4205-ae1a-2ebb9006f9d9)

sample:
[mask.json](https://github.com/user-attachments/files/18181394/mask.json)